File: server/internal/game/physics.go
package game

import (
	"log"
	"math"
	"time"

	"github.com/race/server/config"
)

// Physics handles all physics calculations
type Physics struct{}

// NewPhysics creates a new physics engine
func NewPhysics() *Physics {
	return &Physics{}
}

// UpdatePlayer updates a single player's physics state
func (ph *Physics) UpdatePlayer(p *Player, dt float64) {
	p.mu.Lock()
	defer p.mu.Unlock()

	if p.Exploded {
		return
	}

	input := p.CurrentInput

	// Decode input
	accForce := 0.0
	turnDir := 0.0

	// From keys (bit flags)
	if input.Keys&1 != 0 { // Up
		accForce = config.Acceleration
	}
	if input.Keys&2 != 0 { // Down
		accForce = -config.Braking
	}
	if input.Keys&4 != 0 { // Left
		turnDir = -1.0
	}
	if input.Keys&8 != 0 { // Right
		turnDir = 1.0
	}

	// From analog input (overrides keys if present)
	if math.Abs(input.Throttle) > 0.1 {
		if input.Throttle > 0 {
			accForce = config.Acceleration * input.Throttle
		} else {
			accForce = config.Braking * input.Throttle
		}
	}
	if math.Abs(input.Steering) > 0.1 {
		turnDir = input.Steering
	}

	// Check road boundaries
	roadCenter := config.GetRoadCurve(p.Y)
	distFromCenter := math.Abs(p.X - roadCenter)
	roadHalfWidth := config.RoadWidth / 2.0
	carHalfWidth := config.CarWidth / 2.0
	edgeDist := distFromCenter - roadHalfWidth
	isOffRoad := edgeDist > -carHalfWidth

	// Explosion check
	if edgeDist > config.RoadWidth*config.ExplosionTolerance {
		if !p.Exploded {
			p.Exploded = true
			p.Rating = 0
			p.ExplodedAt = time.Now()
			log.Printf("Player %d exploded: X=%.0f, roadCenter=%.0f, edgeDist=%.0f", p.ID, p.X, roadCenter, edgeDist)
		}
		return
	}

	// Friction
	var activeFriction float64
	if isOffRoad {
		activeFriction = config.FrictionOffroad
	} else {
		activeFriction = config.FrictionRoad
	}

	// Apply friction when not accelerating
	if accForce == 0 {
		if p.Speed > 0 {
			p.Speed = math.Max(0, p.Speed-activeFriction*dt)
		} else if p.Speed < 0 {
			p.Speed = math.Min(0, p.Speed+activeFriction*dt)
		}
	}

	// Off-road speed reduction
	if isOffRoad && accForce != 0 {
		p.Speed -= p.Speed * 2.0 * dt
	}

	// Apply acceleration
	p.Speed += accForce * dt
	p.Speed = math.Max(-config.MaxSpeed*0.2, math.Min(p.Speed, config.MaxSpeed))

	// Steering with understeer
	speedRatio := math.Abs(p.Speed) / config.MaxSpeed
	understeerFactor := math.Max(config.MinTurnAuthority, 1.0-(speedRatio*config.InertiaDampening))

	if math.Abs(turnDir) > 0.01 && math.Abs(p.Speed) > 20 {
		p.X += turnDir * config.TurnSpeed * understeerFactor * dt
		p.Angle = turnDir * 25.0 * understeerFactor

		// Speed penalty from turning
		p.Speed *= 1.0 - (0.3 * math.Abs(turnDir) * dt)
	} else {
		p.Angle *= 0.9
	}

	// Update position
	p.Y += p.Speed * dt

	// Update rating
	if p.Speed > 0 {
		speedFactor := p.Speed / 100.0
		p.Rating += (speedFactor * speedFactor) * dt * 0.5
	}


}

// CheckCollision checks and resolves collision between two players
func (ph *Physics) CheckCollision(p1, p2 *Player, dt float64) bool {
	p1.mu.Lock()
	p2.mu.RLock()

	dx := p1.X - p2.X
	dy := p1.Y - p2.Y
	dist := math.Sqrt(dx*dx + dy*dy)
	minDist := config.CollisionRadius

	if dist >= minDist || dist == 0 {
		p1.mu.Unlock()
		p2.mu.RUnlock()
		return false
	}

	// Normalize collision vector
	nx := dx / dist
	ny := dy / dist
	otherSpeed := p2.Speed
	speedDiff := p1.Speed - otherSpeed

	pushPower := config.PushForce * (math.Abs(p1.Speed) + 100) * dt

	// Speed differential amplification
	if speedDiff > config.SpeedDiffThreshold {
		pushPower *= config.SpeedDiffMultiplier
	}

	p1.X += nx * pushPower
	p1.Y += ny * pushPower
	p1.Speed *= 0.9

	p1.mu.Unlock()
	p2.mu.RUnlock()

	return true
}

// Distance calculates distance between two points
func Distance(x1, y1, x2, y2 float64) float64 {
	dx := x2 - x1
	dy := y2 - y1
	return math.Sqrt(dx*dx + dy*dy)
}

File: client/src/game/physics.ts
import { CONFIG, getRoadCurve } from '@/config';
import { GameStateManager } from './state';
import { Particle } from '@/types';

export class Physics {
  private stateManager: GameStateManager;

  constructor(stateManager: GameStateManager) {
    this.stateManager = stateManager;
  }

  // Update local player physics
  update(dt: number, canvas: HTMLCanvasElement): void {
    const state = this.stateManager.gameState;
    const p = state.localPlayer;

    if (p.exploded) return;

    const { keys, mouse, controlMode } = state;

    let accForce = 0;
    let turnDir = 0;

    // Update rating based on speed
    if (p.speed > 0) {
      const speedFactor = p.speed / 100;
      p.rating += (speedFactor * speedFactor) * dt * 0.5;
    }

    // Process input based on control mode
    if (controlMode === 'keyboard') {
      if (keys.ArrowUp) accForce = CONFIG.ACCELERATION;
      if (keys.ArrowDown) accForce = -CONFIG.BRAKING;
      if (keys.ArrowLeft) turnDir = -1;
      if (keys.ArrowRight) turnDir = 1;
    } else {
      // Mouse control
      const cx = canvas.width / 2;
      const cy = canvas.height * CONFIG.CAMERA_Y_OFFSET;
      const dx = mouse.x - cx;
      const dy = mouse.y - cy;

      const turnRatio = Math.min(1, Math.max(-1, dx / CONFIG.MOUSE_SENSITIVITY_X));
      turnDir = Math.abs(turnRatio) > 0.1 ? turnRatio : 0;

      if (dy < -20) {
        const gasRatio = Math.min(1, Math.abs(dy) / CONFIG.MOUSE_SENSITIVITY_Y);
        accForce = CONFIG.ACCELERATION * gasRatio;
      } else if (dy > 20) {
        accForce = -CONFIG.BRAKING;
      }
    }

    // Check road boundaries
    const roadCenter = getRoadCurve(p.y);
    const distFromCenter = Math.abs(p.x - roadCenter);
    const roadHalfWidth = CONFIG.ROAD_WIDTH / 2;
    const carHalfWidth = CONFIG.CAR_WIDTH / 2;
    const edgeDist = distFromCenter - roadHalfWidth;
    const isOffRoad = edgeDist > -carHalfWidth;

    // Explosion check
    if (edgeDist > CONFIG.ROAD_WIDTH * CONFIG.EXPLOSION_TOLERANCE) {
      this.triggerExplosion();
      return;
    }

    // Friction
    const activeFriction = isOffRoad ? CONFIG.FRICTION_OFFROAD : CONFIG.FRICTION_ROAD;

    // Natural friction decay
    if (accForce === 0 && p.speed > 0) {
      p.speed = Math.max(0, p.speed - activeFriction * dt);
    }
    if (accForce === 0 && p.speed < 0) {
      p.speed = Math.min(0, p.speed + activeFriction * dt);
    }

    // Off-road speed reduction
    if (isOffRoad && accForce !== 0) {
      p.speed -= p.speed * 2.0 * dt;
    }

    // Apply acceleration
    p.speed += accForce * dt;
    p.speed = Math.max(-CONFIG.MAX_SPEED * 0.2, Math.min(p.speed, CONFIG.MAX_SPEED));

    // Steering with understeer
    const speedRatio = Math.abs(p.speed) / CONFIG.MAX_SPEED;
    const understeerFactor = Math.max(CONFIG.MIN_TURN_AUTHORITY, 1.0 - (speedRatio * CONFIG.INERTIA_DAMPENING));

    if (Math.abs(turnDir) > 0.01 && Math.abs(p.speed) > 20) {
      p.x += turnDir * CONFIG.TURN_SPEED * understeerFactor * dt;
      p.angle = turnDir * 25 * understeerFactor;

      // Speed penalty from turning
      p.speed *= 1 - (0.3 * Math.abs(turnDir) * dt);
    } else {
      p.angle *= 0.9;
    }

    // Update position
    p.y += p.speed * dt;

    // Check collisions with remote players
    this.checkCollisions(dt);

    // Decay camera shake
    this.stateManager.decayCameraShake();
  }

  // Check collisions with remote players
  private checkCollisions(dt: number): void {
    const p = this.stateManager.localPlayer;

    this.stateManager.remotePlayers.forEach((other) => {
      const dx = p.x - other.currentX;
      const dy = p.y - other.currentY;
      const dist = Math.sqrt(dx * dx + dy * dy);
      const minDist = CONFIG.CAR_WIDTH * 1.4;

      if (dist < minDist && dist > 0) {
        const nx = dx / dist;
        const ny = dy / dist;
        const otherSpeed = other.speed || 0;
        const speedDiff = p.speed - otherSpeed;

        let pushPower = CONFIG.PUSH_FORCE * (Math.abs(p.speed) + 100) * dt;

        if (speedDiff > CONFIG.SPEED_DIFF_THRESHOLD) {
          pushPower *= CONFIG.SPEED_DIFF_MULTIPLIER;
          this.stateManager.shakeCamera(
            Math.random() * 10 - 5,
            Math.random() * 10 - 5
          );
        }

        p.x += nx * pushPower;
        p.y += ny * pushPower;
        p.speed *= 0.9;
      }
    });
  }

  // Update remote player positions using dead reckoning
  updateRemotePlayers(): void {
    const now = Date.now();

    this.stateManager.remotePlayers.forEach((remote) => {
      // Predict future position based on last packet
      const timeSincePacket = (now - remote.lastPacketTime) / 1000;
      const predictedY = remote.packetY + (remote.speed * timeSincePacket);
      const predictedX = remote.packetX;

      // Interpolate towards prediction
      const t = 0.1;
      remote.currentX = remote.currentX + (predictedX - remote.currentX) * t;
      remote.currentY = remote.currentY + (predictedY - remote.currentY) * t;
    });
  }

  // Trigger player explosion
  private triggerExplosion(): void {
    const p = this.stateManager.localPlayer;
    if (p.exploded) return;

    // Spawn explosion particles
    const particles = this.createExplosionParticles(p.x, p.y, p.color);
    this.stateManager.addParticles(particles);

    // Explode player
    this.stateManager.explodePlayer();

    // Switch to keyboard/joystick mode for safety (not mouse/tilt)
    const mode = this.stateManager.controlMode;
    if (mode === 'mouse' || mode === 'tilt') {
      this.stateManager.forceDefaultMode(mode === 'tilt');
    }
  }

  // Create explosion particles
  private createExplosionParticles(x: number, y: number, color: string): Particle[] {
    const particles: Particle[] = [];

    for (let i = 0; i < CONFIG.EXPLOSION_PARTICLES; i++) {
      particles.push({
        x,
        y,
        vx: (Math.random() - 0.5) * 800,
        vy: (Math.random() - 0.5) * 800,
        life: 1.0,
        color: i % 2 === 0 ? color : '#fbbf24',
        size: Math.random() * 6 + 2,
      });
    }

    return particles;
  }

  // Update particles
  updateParticles(dt: number): void {
    const { particles } = this.stateManager.gameState;

    for (let i = particles.length - 1; i >= 0; i--) {
      const p = particles[i];
      p.x += p.vx * dt;
      p.y += p.vy * dt;
      p.life -= dt * CONFIG.PARTICLE_DECAY;

      if (p.life <= 0) {
        particles.splice(i, 1);
      }
    }
  }

  // Predict upcoming turns
  predictTurn(): { isSharp: boolean; direction: 'left' | 'right' | null } {
    const p = this.stateManager.localPlayer;
    const currentX = getRoadCurve(p.y);
    const futureX = getRoadCurve(p.y + CONFIG.TURN_LOOKAHEAD);
    const delta = futureX - currentX;

    if (Math.abs(delta) > CONFIG.SHARP_TURN_THRESHOLD) {
      return {
        isSharp: true,
        direction: delta > 0 ? 'right' : 'left',
      };
    }

    return { isSharp: false, direction: null };
  }
}
